action_server/src/robocorp/action_server/_server_expose.py

```diff
@@ -107,119 +131,176 @@ async def expose_server(
     Exposes the server to the world.
     """
 
+    pong_queue: asyncio.Queue = asyncio.Queue(maxsize=1)
+    no_connection_queue: asyncio.Queue = asyncio.Queue(maxsize=1)
+
     async def listen_for_requests() -> None:
         max_retries = 3
-        retry_delay = 1
+        retry_delay = 3
         retries = 0
 
         session_payload: Optional[SessionPayload] = (
             get_expose_session_payload(expose_session) if expose_session else None
         )
         while retries < max_retries:
-            try:
-                headers = (
-                    {
-                        "x-session-id": session_payload.sessionId,
-                        "x-session-secret": session_payload.sessionSecret,
-                    }
-                    if session_payload
-                    else {}
-                )
+            headers = (
+                {
+                    "x-session-id": session_payload.sessionId,
+                    "x-session-secret": session_payload.sessionSecret,
+                }
+                if session_payload
+                else {}
+            )
 
+            try:
                 async with websockets.connect(
                     f"wss://client.{expose_url}",
                     extra_headers=headers,
                     logger=log,
+                    close_timeout=0,
                 ) as ws:
-                    while True:
-                        message = await ws.recv()
+                    if retries > 0:
+                        log.info("Reconnected to expose tunnel server")
+                        retries = 0
+
+                    ping_task = asyncio.create_task(
+                        handle_ping_pong(ws, pong_queue, no_connection_queue)
+                    )
+
+                    try:
+                        while True:
+                            done, pending = await asyncio.wait(
```

Comment:
Heartbeats are running in a separate loop, so I'm using the `no_connection_queue` to pass a signal that we have lost connection to the tunnel.